### PR TITLE
Language Independence

### PR DESCRIPTION
**Pull request by kushalsharma**

### CODE REVIEW

File: lyrics-android/pom.xml [hunk 1]
```diff
@@ -19,7 +19,8 @@
     <dependencies>
         <dependency>
             <groupId>com.flipkart.rome</groupId>
-            <artifactId>lyrics-core</artifactId>
+            <artifactId>lyrics-java</artifactId>
+            <version>0.4.0</version>
```
Comment (shrey-garg) [inline lyrics-android/pom.xml]: version not required

File: lyrics-core/pom.xml [hunk 2]
```diff
@@ -49,6 +44,7 @@
         <dependency>
             <groupId>org.junit.jupiter</groupId>
             <artifactId>junit-jupiter-api</artifactId>
+            <version>RELEASE</version>
```
Comment (shrey-garg) [inline lyrics-core/pom.xml]: version not required

File: lyrics-core/src/main/java/com/flipkart/lyrics/interfaces/AnnotationSpec.java [hunk 3]
```diff
@@ -0,0 +1,56 @@
...
+
+import java.util.ArrayList;
+import java.util.List;
+
+/**
+ * @author kushal.sharma on 10/08/17.
+ */
+public class AnnotationSpec {
+    public final Class<?> clazz;
+    public final ClassName className;
```
Comment (shrey-garg) [inline lyrics-core/src/main/java/com/flipkart/lyrics/interfaces/AnnotationSpec.java]: why a ClassName instead of a TypeName?

File: lyrics-core/src/main/java/com/flipkart/lyrics/interfaces/AnnotationSpec.java [hunk 4]
```diff
@@ -0,0 +1,56 @@
...
+    public final List<NameFormatArgs> members = new ArrayList<>();
+
+    public AnnotationSpec(Builder builder) {
+        this.clazz = builder.clazz;
+        this.className = builder.className;
+        this.members.addAll(builder.members);
+    }
+
+    public static Builder builder(Class<?> clazz) {
+        return Song.factory.createAnnotationBuilder(clazz);
```
Comment (shrey-garg) [inline lyrics-core/src/main/java/com/flipkart/lyrics/interfaces/AnnotationSpec.java]: create a ClassName out of the Class, will reduce the number of methods to override in the factory

File: lyrics-core/src/main/java/com/flipkart/lyrics/interfaces/AnnotationSpec.java [hunk 5]
```diff
@@ -0,0 +1,56 @@
...
+    }
+
+    public static Builder builder(ClassName className) {
+        return Song.factory.createAnnotationBuilder(className);
+    }
+
+    public Object getAnnotationSpec() {
+        return null;
```
Comment (shrey-garg) [inline lyrics-core/src/main/java/com/flipkart/lyrics/interfaces/AnnotationSpec.java]: null?

File: lyrics-core/src/main/java/com/flipkart/lyrics/interfaces/AnnotationSpec.java [hunk 6]
```diff
@@ -0,0 +1,56 @@
...
+import java.util.ArrayList;
+import java.util.List;
+
+/**
+ * @author kushal.sharma on 10/08/17.
+ */
+public class AnnotationSpec {
+    public final Class<?> clazz;
+    public final ClassName className;
+    public final List<NameFormatArgs> members = new ArrayList<>();
```
Comment (shrey-garg) [inline lyrics-core/src/main/java/com/flipkart/lyrics/interfaces/AnnotationSpec.java]: Lets discuss this

File: lyrics-core/src/main/java/com/flipkart/lyrics/interfaces/AnnotationSpec.java [hunk 7]
```diff
@@ -0,0 +1,56 @@
...
+        return Song.factory.createAnnotationBuilder(className);
+    }
+
+    public Object getAnnotationSpec() {
+        return null;
+    }
+
+    public static abstract class Builder {
+        private Class<?> clazz;
+        private ClassName className;
```
Comment (shrey-garg) [inline lyrics-core/src/main/java/com/flipkart/lyrics/interfaces/AnnotationSpec.java]: TypeName?

File: lyrics-core/src/main/java/com/flipkart/lyrics/interfaces/AnnotationSpec.java [hunk 8]
```diff
@@ -0,0 +1,56 @@
...
+        private final List<NameFormatArgs> members = new ArrayList<>();
+
+        protected Builder(Class<?> clazz) {
+            this.clazz = clazz;
```
Comment (shrey-garg) [inline lyrics-core/src/main/java/com/flipkart/lyrics/interfaces/AnnotationSpec.java]: lets restrict to just one constructor as we can always create a ClassName out of a class

File: lyrics-android/src/main/java/com/flipkart/lyrics/android/annotators/validations/AndroidValidationStyle.java [hunk 9]
```diff
@@ -17,10 +17,10 @@
 package com.flipkart.lyrics.android.annotators.validations;
 
 import com.flipkart.lyrics.annotators.validations.ValidationAnnotatorStyle;
```
Comment (namannigam) [inline lyrics-android/src/main/java/com/flipkart/lyrics/android/annotators/validations/AndroidValidationStyle.java]: @kushalsharma @shrey-garg Are we planning to move away from javapoet and implement similar generation in lyrics? Any add ons there? **Curious**

**Comment by namannigam** [comment]
@kushalsharma @shrey-garg Are we planning to move away from javapoet and implement similar generation in lyrics? Any add ons there? _Curious_

File: lyrics-core/src/main/java/com/flipkart/lyrics/specs/AnnotationSpec.java [hunk 10]
```diff
@@ -0,0 +1,103 @@
...
+
+    public static Builder builder(ClassName className) {
+        return Song.factory.createAnnotationBuilder(className);
+    }
+
+    public Object getAnnotationSpec() {
+        return null;
+    }
+
+    void emit(CodeWriter codeWriter, boolean inline) throws IOException {
```
Comment (shrey-garg) [inline lyrics-core/src/main/java/com/flipkart/lyrics/specs/AnnotationSpec.java]: is this used anywhere?

File: lyrics-core/src/main/java/com/flipkart/lyrics/specs/AnnotationSpec.java [hunk 11]
```diff
@@ -0,0 +1,103 @@
...
+
+    public static Builder builder(Class<?> clazz) {
+        return Song.factory.createAnnotationBuilder(ClassName.get(clazz));
+    }
+
+    public static Builder builder(ClassName className) {
+        return Song.factory.createAnnotationBuilder(className);
+    }
+
+    public Object getAnnotationSpec() {
```
Comment (shrey-garg) [inline lyrics-core/src/main/java/com/flipkart/lyrics/specs/AnnotationSpec.java]: used anywhere?

File: lyrics-core/src/main/java/com/flipkart/lyrics/specs/AnnotationSpec.java [hunk 12]
```diff
@@ -0,0 +1,103 @@
...
+
+import com.flipkart.lyrics.Song;
+
+import java.io.IOException;
+import java.util.*;
+
+/**
+ * @author kushal.sharma on 10/08/17.
+ */
+public class AnnotationSpec {
```
Comment (shrey-garg) [inline lyrics-core/src/main/java/com/flipkart/lyrics/specs/AnnotationSpec.java]: add a toBuilder method

File: lyrics-core/src/main/java/com/flipkart/lyrics/helper/Util.java [hunk 13]
```diff
@@ -0,0 +1,145 @@
...
+    public static void checkState(boolean condition, String format, Object... args) {
+        if (!condition) {
+            throw new IllegalStateException(String.format(format, args));
+        }
+    }
+
+    /**
+     * Returns the string literal representing {@code value}, including wrapping double quotes.
+     */
+    public static String stringLiteralWithDoubleQuotes(String value, String indent) {
```
Comment (shrey-garg) [inline lyrics-core/src/main/java/com/flipkart/lyrics/helper/Util.java]: unused

File: lyrics-core/src/main/java/com/flipkart/lyrics/helper/Util.java [hunk 14]
```diff
@@ -0,0 +1,145 @@
...
+            // need to append indent after linefeed?
+            if (c == '\n' && i + 1 < value.length()) {
+                result.append("\"\n").append(indent).append(indent).append("+ \"");
+            }
+        }
+        result.append('"');
+        return result.toString();
+    }
+
+    static String characterLiteralWithoutSingleQuotes(char c) {
```
Comment (shrey-garg) [inline lyrics-core/src/main/java/com/flipkart/lyrics/helper/Util.java]: unused

File: lyrics-core/src/main/java/com/flipkart/lyrics/specs/AnnotationSpec.java [hunk 15]
```diff
@@ -0,0 +1,107 @@
...
+
+    public Builder toBuilder() {
+        Builder builder = new Builder(type);
+        for (Map.Entry<String, List<CodeBlock>> entry : members.entrySet()) {
+            builder.members.put(entry.getKey(), new ArrayList<>(entry.getValue()));
+        }
+        return builder;
+    }
+
+    void emit(CodeWriter codeWriter, boolean inline) throws IOException {
```
Comment (shrey-garg) [inline lyrics-core/src/main/java/com/flipkart/lyrics/specs/AnnotationSpec.java]: lets remove these emit methods

File: lyrics-core/src/main/java/com/flipkart/lyrics/specs/AnnotationSpec.java [hunk 16]
```diff
@@ -0,0 +1,107 @@
...
+                codeWriter.emit("$L = ", entry.getKey());
+                emitAnnotationValues(codeWriter, whitespace, memberSeparator, entry.getValue());
+                if (i.hasNext()) codeWriter.emit(memberSeparator);
+            }
+            codeWriter.unindent(2);
+            codeWriter.emit(whitespace + ")");
+        }
+    }
+
+    private void emitAnnotationValues(CodeWriter codeWriter, String whitespace,
```
Comment (shrey-garg) [inline lyrics-core/src/main/java/com/flipkart/lyrics/specs/AnnotationSpec.java]: same

File: lyrics-core/src/main/java/com/flipkart/lyrics/specs/AnnotationSpec.java [hunk 17]
```diff
@@ -0,0 +1,107 @@
+package com.flipkart.lyrics.specs;
```
Comment (shrey-garg) [inline lyrics-core/src/main/java/com/flipkart/lyrics/specs/AnnotationSpec.java]: add a copyright notice like other classes in the repo

File: lyrics-core/src/main/java/com/flipkart/lyrics/specs/ArrayTypeName.java [hunk 18]
```diff
@@ -0,0 +1,87 @@
+/*
+ * Copyright (C) 2015 Square, Inc.
```
Comment (shrey-garg) [inline lyrics-core/src/main/java/com/flipkart/lyrics/specs/ArrayTypeName.java]: wrong one, use the one for flipkart

File: lyrics-core/src/main/java/com/flipkart/lyrics/processor/types/TypeAdditionalHandler.java [hunk 19]
```diff
@@ -1,9 +1,23 @@
+/*
+ * Copyright 2016 Flipkart Internet, pvt ltd.
```
Comment (shrey-garg) [inline lyrics-core/src/main/java/com/flipkart/lyrics/processor/types/TypeAdditionalHandler.java]: 2017

File: lyrics-core/src/main/java/com/flipkart/lyrics/specs/ClassName.java [hunk 20]
```diff
@@ -0,0 +1,246 @@
+/*
+ * Copyright (C) 2014 Google, Inc.
```
Comment (shrey-garg) [inline lyrics-core/src/main/java/com/flipkart/lyrics/specs/ClassName.java]: is this coming class coming from google? If so, can we use the artifact that contains this class?

File: lyrics-core/src/test/java/com/flipkart/lyrics/annotators/GsonStyleTest.java [hunk 21]
```diff
@@ -51,13 +48,6 @@ public void testNamedAsRule() {
         FieldSpec fieldSpec = builder.build();
         assertEquals("test", fieldSpec.name);
         assertEquals(2, fieldSpec.annotations.size());
-
-        AnnotationSpec namedAsAnnotation = fieldSpec.annotations.get(0);
-        assertEquals(SerializedName.class.getName(), namedAsAnnotation.type.toString());
-        assertEquals(1, namedAsAnnotation.members.size());
-        assertEquals("\"TEST\"", namedAsAnnotation.members.get("value").get(0).toString());
```
Comment (shrey-garg) [inline lyrics-core/src/test/java/com/flipkart/lyrics/annotators/GsonStyleTest.java]: other than this line, can other still be valid?

File: lyrics-core/src/test/java/com/flipkart/lyrics/processor/modifiers/ModifiersHandlerTest.java [hunk 22]
```diff
@@ -103,8 +103,8 @@ public void testConfigModifiersAndModelModifier(@TuneProvider Tune tune) {
         Set<Modifier> modifiers = spec.modifiers;
 
         assertEquals(2, modifiers.size(), "No Modifiers found");
-        assertTrue(spec.hasModifier(Modifier.PRIVATE));
-        assertTrue(spec.hasModifier(Modifier.ABSTRACT));
+        assertTrue(spec.hasModifier((Modifier.PRIVATE)));
```
Comment (shrey-garg) [inline lyrics-core/src/test/java/com/flipkart/lyrics/processor/modifiers/ModifiersHandlerTest.java]: why the extra brackets?

File: lyrics-core/src/test/java/com/flipkart/lyrics/processor/supertypes/SuperClassHandlerTest.java [hunk 23]
```diff
@@ -53,9 +53,17 @@ public void setUp() {
 
     @Test
     public void testNoSuperClass(@TuneProvider Tune tune) {
-        TypeModel model = new TypeModel();
+        String fullPackage = "java.lang";
```
Comment (shrey-garg) [inline lyrics-core/src/test/java/com/flipkart/lyrics/processor/supertypes/SuperClassHandlerTest.java]: why is this change required?

File: lyrics-samples/lyrics-json-sample/src/main/java/com/flipkart/lyrics/sample/json/Generator.java [hunk 24]
```diff
@@ -1,32 +1,40 @@
...
-import java.nio.file.Paths;
 
 /**
  * Created by shrey.garg on 04/02/17.
  */
 public class Generator {
 
     public static void main(String[] args) {
-        Tune tune = new DefaultTune();
+        Tune tune = new TestTune();
```
Comment (shrey-garg) [inline lyrics-samples/lyrics-json-sample/src/main/java/com/flipkart/lyrics/sample/json/Generator.java]: JavaTune

File: lyrics-samples/lyrics-json-sample/src/main/java/com/flipkart/lyrics/sample/json/Generator.java [hunk 25]
```diff
@@ -1,32 +1,40 @@
...
 
         File target = new File(Generator.class.getProtectionDomain().getCodeSource().getLocation().getPath()).getParentFile();
         File generatedSources = new File(target, "generated-sources");
 
         Lyrics lyrics = new Lyrics(tune, generatedSources);
 
         // The working directory should be lyrics-json-sample
-        File source = new File("src/main/resources/com");
+        File source = new File("lyrics-samples/lyrics-json-sample/src/main/resources/com");
```
Comment (shrey-garg) [inline lyrics-samples/lyrics-json-sample/src/main/java/com/flipkart/lyrics/sample/json/Generator.java]: revert

File: lyrics-core/src/main/java/com/flipkart/lyrics/specs/AnnotationSpec.java (hunk 17 above)
Comment (kushalsharma) [reply lyrics-core/src/main/java/com/flipkart/lyrics/specs/AnnotationSpec.java]: done

File: lyrics-core/src/main/java/com/flipkart/lyrics/specs/ClassName.java (hunk 20 above)
Comment (kushalsharma) [reply lyrics-core/src/main/java/com/flipkart/lyrics/specs/ClassName.java]: Cannot find any artifact with this class. But square guys are licensing this class to Google.

File: lyrics-core/src/main/java/com/flipkart/lyrics/specs/ArrayTypeName.java (hunk 18 above)
Comment (kushalsharma) [reply lyrics-core/src/main/java/com/flipkart/lyrics/specs/ArrayTypeName.java]: as discussed offline.

File: lyrics-core/src/main/java/com/flipkart/lyrics/specs/CodeBlock.java [hunk 26]
```diff
@@ -0,0 +1,113 @@
...
+        public Builder add(String format, Object... args) {
+            this.formats.add(format);
+            for (Object object : args) {
+                this.arguments.add(object);
+            }
+            return this;
+        }
+
+        public Builder beginControlFlow(String controlFlow, Object... args) {
+            add(controlFlow + " {\n", args);
```
Comment (shrey-garg) [inline lyrics-core/src/main/java/com/flipkart/lyrics/specs/CodeBlock.java]: lets discuss this tomorrow

File: lyrics-core/src/main/java/com/flipkart/lyrics/styles/objectmethods/CommonsLang3Style.java [hunk 27]
```diff
@@ -54,14 +53,14 @@ public void processEqualsAndHashCode(MethodSpec.Builder equalsBuilder, MethodSpe
         equalsBuilder.addCode("return new $T()\n", equalsBuilderClass);
 
         if (testSuperEquality) {
-            equalsBuilder.addCode("\t\t.appendSuper(super.equals(that))\n", equalsBuilderClass);
+            equalsBuilder.addCode("\t\t.appendSuper(super.equals(that))\n");
```
Comment (shrey-garg) [inline lyrics-core/src/main/java/com/flipkart/lyrics/styles/objectmethods/CommonsLang3Style.java]: why this change?

File: lyrics-java/pom.xml [hunk 28]
```diff
@@ -0,0 +1,37 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<project xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
+         xmlns="http://maven.apache.org/POM/4.0.0"
+         xsi:schemaLocation="http://maven.apache.org/POM/4.0.0 http://maven.apache.org/xsd/maven-4.0.0.xsd">
+    <modelVersion>4.0.0</modelVersion>
+
+    <groupId>com.flipkart.rome</groupId>
```
Comment (shrey-garg) [inline lyrics-java/pom.xml]: make this a child module

File: lyrics-java/pom.xml [hunk 29]
```diff
@@ -0,0 +1,37 @@
...
+    <artifactId>lyrics-java</artifactId>
+    <version>0.4.0</version>
+    <build>
```
Comment (shrey-garg) [inline lyrics-java/pom.xml]: remove

File: lyrics-java/pom.xml [hunk 30]
```diff
@@ -0,0 +1,37 @@
...
+                </configuration>
+            </plugin>
+        </plugins>
+    </build>
+
+    <dependencies>
+        <dependency>
+            <groupId>com.flipkart.rome</groupId>
+            <artifactId>lyrics-core</artifactId>
+            <version>0.4.0</version>
```
Comment (shrey-garg) [inline lyrics-java/pom.xml]: remember to remove version

File: lyrics-json/pom.xml [hunk 31]
```diff
@@ -19,7 +19,14 @@
...
         <dependency>
             <groupId>com.flipkart.rome</groupId>
-            <artifactId>lyrics-core</artifactId>
+            <artifactId>lyrics-java</artifactId>
+            <version>0.4.0</version>
+        </dependency>
+
+        <dependency>
+            <groupId>com.flipkart.rome</groupId>
+            <artifactId>lyrics-android</artifactId>
```
Comment (shrey-garg) [inline lyrics-json/pom.xml]: we shouldnt need android module as a dependency for json

File: pom.xml [hunk 32]
```diff
@@ -34,18 +35,6 @@
     <dependencyManagement>
         <dependencies>
             <dependency>
-                <groupId>com.flipkart.rome</groupId>
```
Comment (shrey-garg) [inline pom.xml]: revert the changes here
